Unknown computed column type

boutinb opened the pull request:
The computedColumnType 'unknown' does not exist anymore, but might be stored in the database. So we need to take care of this.
Fixes https://github.com/jasp-stats/jasp-issues/issues/2517

boutinb committed:
Unknown computed column type

The computedColumnType 'unknown' does not exist anymore, but might be stored in the database. So we need to take care of this.
Fixes jasp-stats/jasp-issues#2517

boutinb committed:
Use try-catch instead

JorisGoosen commented on CommonData/databaseinterface.cpp:
```diff
@@ -1092,7 +1092,12 @@ bool DatabaseInterface::columnGetComputedInfo(int columnId, int &analysisId, boo
 
-		codeType = codeTypeStr.empty() ? computedColumnType::notComputed : computedColumnTypeFromString(codeTypeStr);
+		codeType = computedColumnType::notComputed;
+		if (!codeTypeStr.empty())
+		{
+			try { codeType = computedColumnTypeFromString(codeTypeStr); }
+			catch(...) {}
+		}
```
```suggestion
		try { codeType = computedColumnTypeFromString(codeTypeStr); }
		catch(...) {}
```

There is no need for the if, if it is empty it will also fail and thus get caught

boutinb replied:
Yes but I saw that in the file that was given in the issue, that most of the columns have empty value. If there are hundreds of columns, that would make hundreds of catch exceptions, which can be a bit long.

JorisGoosen replied:
Ok well I dont think a few hundred exceptions will make a difference but whatever.

JorisGoosen commented on CommonData/column.cpp:
```diff
@@ -65,12 +65,22 @@ void Column::loadComputedColumnJsonBackwardsCompatibly(const Json::Value & json)
 	
-	const std::string & rCode = json["rCode"].asString();
-	
+	const std::string & rCode = json["rCode"].asString(),
+					  & codeTypeStr = json["codeType"].asString();
+
+	computedColumnType codeType = computedColumnType::notComputed;
+	if (!codeTypeStr.empty())
+	{
+		try { codeType = computedColumnTypeFromString(codeTypeStr); }
+		catch(...) {}
+	}
+
+	try { codeType = computedColumnTypeFromString(codeTypeStr); } catch(...) {}
+
 	setCompColStuff
 	(
 		json["invalidated"].asBool(),
-		computedColumnTypeFromString(json["codeType"].asString()),
+		codeType,
 		rCode,
```
You dont need to make changes to column.cpp I think?

Just the change to `DatabaseInterface::columnGetComputedInfo` should be enough

boutinb replied:
I was not quite completely sure that this error could not arise here. For the case, it does not hurt I fimd.

JorisGoosen replied:
Yeah but if the changes are not necessary at all why make them?

JorisGoosen commented:
Ah while testing I found that loading the jasp file from the issue with the current PR gets:
<img width="837" alt="image" src="https://github.com/jasp-stats/jasp-desktop/assets/29062713/6f36e4e9-37eb-48b4-b8fa-6a70ae20234c">

So it is "not computed" yet it is computed according to the columnheader.

So this is not really a fix yet.

JorisGoosen commented:
the question is more, how can `unknown` even be written there? Or was that simply the default when we didnt set it before?
How is it possible that a column was stored with `unknown` as computed column type at all?

This has a rootcause and we need to figure it out how it happened.

Also, to expand this workaround it would perhaps be good to check in `DatabaseInterface::columnGetComputedInfo` whether the `computedColumnType` isnt `notComputed` before returning `true` (because that says: `isComputed`)

boutinb commented:
> the question is more, how can `unknown` even be written there? Or was that simply the default when we didnt set it before? How is it possible that a column was stored with `unknown` as computed column type at all?
> 
> This has a rootcause and we need to figure it out how it happened.

When I open the file in 0.18.1, then I see also a not computed column with an icon telling this is a computed column. So the situation was buggy in 0.18.1. The rcode of this column is also empty.
I have tried to reproduce this in 0.18.1 by undo/redo, copy/paste, but could not find a way to get such a situation.
But the code should be better in 0.18.2! So we can expect that this is not reproducible anyway.

The isComputed column in the database is not needed anymore. So I would just remove it.

JorisGoosen commented:
> The isComputed column in the database is not needed anymore. So I would just remove it.

Ok so remove it in `void DatabaseInterface::upgradeDBFromVersion(Version originalVersion)` 
By adding something to it like:
```
	   if(originalVersion < "0.18.2")
			   runStatements("ALTER TABLE DataSets ADD COLUMN description     TEXT;"                 "\n");

	   if(originalVersion < "0.18.3")
			   runStatements("ALTER TABLE Columns DROP COLUMN isComputed;"                 "\n");

```

And then remove it from all the functions that expect it...

You want this to be a hotfix or a potential place to create more bugs though? ;)

boutinb commented:
Hmm by removing isColumn in the database, we have then a downgrade issue: JASP file made in 0.18.3 won't be readable in 0.18.2 or lower... This is not expected for a minor release. So I will let the isColumn in the database. We can remove it in 0.19

JorisGoosen commented:
Exactly, however, you still need to fix the output of that function then. The best thing would be to make sure that if the computedColumnType is "notComputed" the `isComputed` thing is set FALSE

boutinb committed:
Remove the isColumn property

the _isColumn property of Column class can be deduced by the codeType property, so remove it from the class.
However for downgrade purpose, the isColumn column in the database is kept so that JASP file made in 0.18.3 can be still read in 0.18.2 or 0.18.1.

boutinb commented:
I have removed the _isColumn property of the Column class, as it can be deduced by the codeType property
However for downgrade purpose, the isColumn column in the database is kept so that JASP file made in 0.18.3 can be still read in 0.18.2 or 0.18.1.

JorisGoosen committed:
mention isComputed in update function for later removal
